C/services/common-plugin-interfaces/python/include/python_plugin_common_interface.h

```diff
@@ -10,18 +10,61 @@
  * Author: Massimiliano Pinto, Amandeep Singh Arora
  */
 
-#include <map>
+#include <plugin_manager.h>
+
+#define SHIM_SCRIPT_REL_PATH  "/python/foglamp/plugins/common/shim/"
+#define SHIM_SCRIPT_POSTFIX "_shim"
 
 using namespace std;
 
+/**
+ * This class represents the loaded Python module
+ * with interpreter initialisation flag.
+ * That flag is checked in PluginInterfaceCleanup
+ * before removing Python interpreter.
+ */
+class PythonModule
+{
+	public:
+		PythonModule(PyObject* module,
+			    bool init,
+			    string name,
+			    string type,
+			    PyThreadState* state) :
+			m_module(module),
+			m_init(init),
+			m_name(name),
+			m_type(type),
+			m_tState(state)
+		{
+		};
+
+		~PythonModule()
+		{
+			// Destroy loaded Python module
+			Py_CLEAR(m_module);
+			m_module = NULL;
+		};
+
+	public:
+		PyObject* m_module;
+		bool      m_init;
+		string    m_name;
+		string    m_type;
+		PyThreadState*	m_tState;
+};
+
 extern "C" {
 // This is the map of Python object initialised in each 
 // South, Notification, Filter  plugin interfaces
-static map<string, PyObject*> *pythonModules = new map<string, PyObject *>();;
+static map<string, PythonModule*> *pythonModules = new map<string, PythonModule*>();
+// Map of PLUGIN_HANDLE objects, updated by plugin_init calls
+static map<PLUGIN_HANDLE, PythonModule*> *pythonHandles = new map<PLUGIN_HANDLE, PythonModule*>();
 
 // Global variable gPluginName set by PluginInterfaceInit:
 // it has a different memory address when set/read by
 // PluginInterfaceInit in South, Filter or Notification
+// Only used in plugin_info_fn calls
 static string gPluginName;
```

Comment:
Is it not possible to wrap this in a class? I'm concerned by the sharing semantics for these static variables.

If there really only should be one in any given process then I think wrapping them in a single class would make it much clearer. Maybe something like a PythonPluginManager class.